File: frontend/src/api/endpoints/opportunities.ts
import { api } from "@/api/client";

// ─── SuiteCRM Opportunities (OPP-002..OPP-006) ───────────────────────────────
// TS interfaces mirror the LIVE Pydantic models in app/models.py.
// Backend router prefix: /ui/sales  (admin: /ui/admin/sales)
// Feature flag: S.sales_pipeline_enabled → endpoints 503 when off
//               (feature-status never 503s).

// ─── Constants (mirror app/models.py + app/services/opportunities.py) ─────────

export const OPPORTUNITY_STAGES = [
  "prospecting",
  "qualification",
  "needs_analysis",
  "value_proposition",
  "id_decision_makers",
  "proposal_price_quote",
  "negotiation_review",
  "closed_won",
  "closed_lost",
] as const;
export type OpportunityStage = (typeof OPPORTUNITY_STAGES)[number];

export const LEAD_SOURCE_CHOICES = [
  "cold_call",
  "existing_customer",
  "self_generated",
  "employee",
  "partner",
  "public_relations",
  "direct_mail",
  "conference",
  "trade_show",
  "web_site",
  "word_of_mouth",
  "email",
  "campaign",
  "other",
] as const;
export type LeadSource = (typeof LEAD_SOURCE_CHOICES)[number];

export const CONTACT_ROLES = [
  "decision_maker",
  "evaluator",
  "influencer",
  "champion",
  "end_user",
  "executive_sponsor",
  "technical_buyer",
  "other",
] as const;
export type ContactRole = (typeof CONTACT_ROLES)[number];

export const PERIOD_TYPE_CHOICES = ["monthly", "quarterly", "annual"] as const;
export type PeriodType = (typeof PERIOD_TYPE_CHOICES)[number];

// ─── Model interfaces ─────────────────────────────────────────────────────────

export interface OppContactRoleOut {
  opp_id: string;
  contact_ref: string;
  contact_role: string;
  owner_sub: string;
  created_at: number;
}

export interface OppContactRoleIn {
  contact_ref: string;
  contact_role: ContactRole | string;
}

export interface OpportunityOut {
  opp_id: string;
  owner_sub: string;
  name: string;
  stage: string;
  amount_cents: number;
  weighted_amount_cents: number;
  close_date: number;
  probability: number;
  lead_source?: string | null;
  description?: string | null;
  account_party_id?: string | null;
  contact_party_id?: string | null;
  created_at: number;
  updated_at: number;
  contact_roles: OppContactRoleOut[];
}

export interface OpportunityCreateIn {
  name: string;
  stage: OpportunityStage | string;
  amount_cents: number;
  close_date: number;
  probability?: number;
  lead_source?: string | null;
  description?: string | null;
  account_party_id?: string | null;
  contact_party_id?: string | null;
}

export interface OpportunityUpdateIn {
  name?: string;
  stage?: OpportunityStage | string;
  amount_cents?: number;
  close_date?: number;
  probability?: number;
  lead_source?: string | null;
  description?: string | null;
  account_party_id?: string | null;
  contact_party_id?: string | null;
}

export interface OpportunityListResp {
  items: OpportunityOut[];
  next_cursor?: string | null;
}

export interface StageConfigItemOut {
  stage_key: string;
  label: string;
  probability_default: number;
  order: number;
  is_won: boolean;
  is_lost: boolean;
  color?: string | null;
}

export interface StageConfigOut {
  stages: StageConfigItemOut[];
  updated_at?: number | null;
  updated_by_sub?: string | null;
}

export interface StageConfigItemIn {
  stage_key: string;
  label: string;
  probability_default?: number;
  order?: number;
  is_won?: boolean;
  is_lost?: boolean;
  color?: string | null;
}

export interface StageConfigIn {
  stages: StageConfigItemIn[];
}

export interface PipelineStageMetric {
  stage: string;
  label: string;
  count: number;
  total_amount_cents: number;
  weighted_amount_cents: number;
  avg_close_date?: number | null;
}

export interface PipelineReportOut {
  stages: PipelineStageMetric[];
  total_amount_cents: number;
  total_weighted_cents: number;
  generated_at: number;
}

export interface ForecastWorksheetIn {
  committed_cents: number;
  best_case_cents: number;
  pipeline_cents: number;
  notes?: string | null;
}

export interface ForecastWorksheetOut {
  user_sub: string;
  period_key: string;
  committed_cents: number;
  best_case_cents: number;
  pipeline_cents: number;
  closed_cents: number;
  quota_cents: number;
  attainment_pct: number;
  notes?: string | null;
  created_at: number;
  updated_at: number;
}

export interface SalesQuotaIn {
  user_sub: string;
  period_type: PeriodType | string;
  period_key: string;
  target_amount_cents: number;
}

export interface SalesQuotaOut {
  user_sub: string;
  period_type: string;
  period_key: string;
  target_amount_cents: number;
  created_at: number;
  updated_at: number;
  set_by_sub: string;
}

export interface SalesQuotaListOut {
  items: SalesQuotaOut[];
  next_cursor?: string | null;
}

export interface FeatureStatus {
  enabled: boolean;
}

// ─── Feature status (OPP-003/006) ─────────────────────────────────────────────

export const getFeatureStatus = () =>
  api.get<FeatureStatus>("/ui/sales/feature-status");

// ─── OPP-002: Opportunity CRUD ────────────────────────────────────────────────

export const createOpportunity = (body: OpportunityCreateIn) =>
  api.post<OpportunityOut>("/ui/sales/opportunities", body);

export const listOpportunities = (opts?: {
  stage?: string;
  cursor?: string;
  limit?: number;
}) => {
  const params: Record<string, string> = {};
  if (opts?.stage) params["stage"] = opts.stage;
  if (opts?.cursor) params["cursor"] = opts.cursor;
  if (opts?.limit) params["limit"] = String(opts.limit);
  return api.get<OpportunityListResp>("/ui/sales/opportunities", params);
};

export const getOpportunity = (oppId: string, includeContacts = false) =>
  api.get<OpportunityOut>(
    `/ui/sales/opportunities/${encodeURIComponent(oppId)}`,
    includeContacts ? { include_contacts: "true" } : undefined,
  );

export const updateOpportunity = (oppId: string, body: OpportunityUpdateIn) =>
  api.patch<OpportunityOut>(
    `/ui/sales/opportunities/${encodeURIComponent(oppId)}`,
    body,
  );

export const deleteOpportunity = (oppId: string) =>
  api.del<void>(`/ui/sales/opportunities/${encodeURIComponent(oppId)}`);

// ─── OPP-004: Contact-role junction ──────────────────────────────────────────

export const addContactRole = (oppId: string, body: OppContactRoleIn) =>
  api.post<OppContactRoleOut>(
    `/ui/sales/opportunities/${encodeURIComponent(oppId)}/contacts`,
    body,
  );

export const listContactRoles = (oppId: string) =>
  api.get<OppContactRoleOut[]>(
    `/ui/sales/opportunities/${encodeURIComponent(oppId)}/contacts`,
  );

export const removeContactRole = (oppId: string, contactRef: string) =>
  api.del<void>(
    `/ui/sales/opportunities/${encodeURIComponent(oppId)}/contacts/${encodeURIComponent(contactRef)}`,
  );

// ─── OPP-003: Stage config ────────────────────────────────────────────────────

export const listStages = () =>
  api.get<StageConfigOut>("/ui/sales/stages");

export const updateStages = (body: StageConfigIn) =>
  api.put<StageConfigOut>("/ui/admin/sales/stages", body);

// ─── OPP-005: Forecast worksheet ──────────────────────────────────────────────

export const getForecast = (periodKey: string) =>
  api.get<ForecastWorksheetOut>(
    `/ui/sales/forecast/${encodeURIComponent(periodKey)}`,
  );

export const upsertForecast = (periodKey: string, body: ForecastWorksheetIn) =>
  api.put<ForecastWorksheetOut>(
    `/ui/sales/forecast/${encodeURIComponent(periodKey)}`,
    body,
  );

// ─── OPP-006: Pipeline report ─────────────────────────────────────────────────

export const getPipelineReport = (opts?: { from_ts?: number; to_ts?: number }) => {
  const params: Record<string, string> = {};
  if (opts?.from_ts != null) params["from_ts"] = String(opts.from_ts);
  if (opts?.to_ts != null) params["to_ts"] = String(opts.to_ts);
  return api.get<PipelineReportOut>("/ui/sales/reports/pipeline", params);
};

export const getAdminPipelineReport = (opts?: {
  from_ts?: number;
  to_ts?: number;
}) => {
  const params: Record<string, string> = {};
  if (opts?.from_ts != null) params["from_ts"] = String(opts.from_ts);
  if (opts?.to_ts != null) params["to_ts"] = String(opts.to_ts);
  return api.get<PipelineReportOut>("/ui/admin/sales/reports/pipeline", params);
};

// ─── OPP-005: Admin quota ─────────────────────────────────────────────────────

export const setQuota = (body: SalesQuotaIn) =>
  api.post<SalesQuotaOut>("/ui/admin/sales/quotas", body);

export const listUserQuotas = (
  userSub: string,
  opts?: { cursor?: string; limit?: number },
) => {
  const params: Record<string, string> = {};
  if (opts?.cursor) params["cursor"] = opts.cursor;
  if (opts?.limit) params["limit"] = String(opts.limit);
  return api.get<SalesQuotaListOut>(
    `/ui/admin/sales/quotas/${encodeURIComponent(userSub)}`,
    params,
  );
};

// ─── Display helpers ──────────────────────────────────────────────────────────

export const STAGE_LABELS: Record<string, string> = {
  prospecting: "Prospecting",
  qualification: "Qualification",
  needs_analysis: "Needs Analysis",
  value_proposition: "Value Proposition",
  id_decision_makers: "Decision Makers",
  proposal_price_quote: "Proposal / Quote",
  negotiation_review: "Negotiation",
  closed_won: "Closed Won",
  closed_lost: "Closed Lost",
};

export function stageLabel(key: string): string {
  return STAGE_LABELS[key] ?? key.replace(/_/g, " ");
}

export function isWonStage(key: string): boolean {
  return key === "closed_won";
}

export function isLostStage(key: string): boolean {
  return key === "closed_lost";
}

export function isClosedStage(key: string): boolean {
  return isWonStage(key) || isLostStage(key);
}

/** Integer cents → "$1,234.56" */
export function formatCents(cents: number, currency = "USD"): string {
  return new Intl.NumberFormat("en-US", {
    style: "currency",
    currency,
  }).format((cents ?? 0) / 100);
}

export function formatDate(ts?: number | null): string {
  if (!ts) return "—";
  return new Date(ts * 1000).toLocaleDateString();
}

File: app/routers/opportunities.py
"""Sales pipeline — Opportunities router (OPP-002..OPP-006)."""
from __future__ import annotations

from typing import Dict, List, Optional

from fastapi import APIRouter, Depends, HTTPException, Query, Response

from app.auth.roles import Role, normalize_role
from app.auth.policy import require_admin_or_root
from app.auth.deps import get_authenticated_user, AuthenticatedUser
from app.models import (
    ForecastWorksheetIn,
    ForecastWorksheetOut,
    OppContactRoleIn,
    OppContactRoleOut,
    OpportunityCreateIn,
    OpportunityOut,
    OpportunityUpdateIn,
    PipelineReportOut,
    SalesQuotaIn,
    SalesQuotaListOut,
    SalesQuotaOut,
    StageConfigIn,
    StageConfigOut,
)
from app.services.sessions import require_ui_session
from app.services.opportunities import (
    _require_flag,
    add_contact_role,
    create_opportunity,
    delete_opportunity,
    get_forecast,
    get_opportunity,
    get_quota,
    get_stage_config,
    list_contact_roles,
    list_opportunities,
    list_quotas_for_user,
    pipeline_report,
    pipeline_report_all,
    remove_contact_role,
    set_quota,
    set_stage_config,
    update_opportunity,
    upsert_forecast,
)
from app.core.settings import S

router = APIRouter(prefix="/ui/sales", tags=["sales-pipeline"])
router_admin = APIRouter(prefix="/ui/admin/sales", tags=["sales-pipeline-admin"])


# ── Feature status (OPP-003/006) — must not call _require_flag ────────────────

@router.get("/feature-status")
async def feature_status(ctx: dict = Depends(require_ui_session)) -> dict:
    """Return whether the sales pipeline feature is enabled. Does not 503 when off."""
    return {"enabled": S.sales_pipeline_enabled}


# ── OPP-003: Stage config endpoints ──────────────────────────────────────────
# NOTE: declared BEFORE /{opp_id} to avoid FastAPI capturing "stages" as opp_id

@router.get("/stages", response_model=StageConfigOut)
async def list_stages(ctx: dict = Depends(require_ui_session)) -> StageConfigOut:
    """Return active stage list. Requires a valid session; no admin role needed."""
    return get_stage_config()


@router_admin.put("/stages", response_model=StageConfigOut)
async def update_stages(
    body: StageConfigIn,
    ctx: dict = Depends(require_ui_session),
) -> StageConfigOut:
    """Replace the stage list. Admin or Root role required."""
    if normalize_role(ctx.get("role")) not in {Role.ADMIN, Role.ROOT}:
        raise HTTPException(403, detail={"code": "admin_role_required"})
    return set_stage_config(ctx["user_sub"], body)


# ── OPP-005: Forecast worksheet endpoints ─────────────────────────────────────
# NOTE: declared BEFORE /{opp_id} to prevent "forecast" captured as opp_id

@router.put("/forecast/{period_key}", response_model=ForecastWorksheetOut)
async def put_forecast(
    period_key: str,
    body: ForecastWorksheetIn,
    user: AuthenticatedUser = Depends(get_authenticated_user),
) -> ForecastWorksheetOut:
    """Upsert a forecast worksheet for the authenticated rep."""
    _require_flag()
    return upsert_forecast(user.sub, period_key, body)


@router.get("/forecast/{period_key}", response_model=ForecastWorksheetOut)
async def read_forecast(
    period_key: str,
    user: AuthenticatedUser = Depends(get_authenticated_user),
) -> ForecastWorksheetOut:
    """Get a forecast worksheet for the authenticated rep."""
    _require_flag()
    return get_forecast(user.sub, period_key)


# ── OPP-006: Pipeline report endpoints ───────────────────────────────────────
# NOTE: declared BEFORE /{opp_id} to prevent "reports" captured as opp_id

@router.get("/reports/pipeline", response_model=PipelineReportOut)
async def get_pipeline_report(
    from_ts: Optional[int] = Query(None),
    to_ts: Optional[int] = Query(None),
    ctx: dict = Depends(require_ui_session),
) -> PipelineReportOut:
    """Per-rep pipeline funnel report."""
    _require_flag()
    return pipeline_report(ctx["user_sub"], from_ts, to_ts)


@router_admin.get("/reports/pipeline", response_model=PipelineReportOut)
async def get_admin_pipeline_report(
    from_ts: Optional[int] = Query(None),
    to_ts: Optional[int] = Query(None),
    user: AuthenticatedUser = Depends(require_admin_or_root),
) -> PipelineReportOut:
    """Admin cross-user pipeline funnel report."""
    _require_flag()
    return pipeline_report_all(user.sub, from_ts, to_ts)


# ── OPP-002: Opportunity CRUD endpoints ───────────────────────────────────────

@router.post("/opportunities", response_model=OpportunityOut, status_code=201)
async def create_opp(
    body: OpportunityCreateIn,
    ctx: dict = Depends(require_ui_session),
    request=None,
) -> OpportunityOut:
    _require_flag()
    return create_opportunity(ctx["user_sub"], body, request)


@router.get("/opportunities", response_model=Dict)
async def list_opps(
    stage: Optional[str] = Query(None),
    cursor: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    ctx: dict = Depends(require_ui_session),
) -> dict:
    _require_flag()
    items, next_cursor = list_opportunities(ctx["user_sub"], stage, cursor, limit)
    return {"items": [i.model_dump() for i in items], "next_cursor": next_cursor}


@router.get("/opportunities/{opp_id}", response_model=OpportunityOut)
async def read_opp(
    opp_id: str,
    include_contacts: bool = Query(default=False),
    ctx: dict = Depends(require_ui_session),
) -> OpportunityOut:
    _require_flag()
    opp = get_opportunity(ctx["user_sub"], opp_id)
    if include_contacts:
        opp.contact_roles = list_contact_roles(ctx["user_sub"], opp_id)
    return opp


@router.patch("/opportunities/{opp_id}", response_model=OpportunityOut)
async def patch_opp(
    opp_id: str,
    body: OpportunityUpdateIn,
    ctx: dict = Depends(require_ui_session),
    request=None,
) -> OpportunityOut:
    _require_flag()
    return update_opportunity(ctx["user_sub"], opp_id, body, request)


@router.delete("/opportunities/{opp_id}", status_code=204, response_class=Response)
def delete_opp(
    opp_id: str,
    ctx: dict = Depends(require_ui_session),
):
    _require_flag()
    delete_opportunity(ctx["user_sub"], opp_id)


# ── OPP-004: Contact-role junction endpoints ──────────────────────────────────

@router.post("/opportunities/{opp_id}/contacts", response_model=OppContactRoleOut, status_code=201)
def create_contact_role(
    opp_id: str,
    body: OppContactRoleIn,
    ctx: dict = Depends(require_ui_session),
) -> OppContactRoleOut:
    _require_flag()
    return add_contact_role(ctx["user_sub"], opp_id, body)


@router.get("/opportunities/{opp_id}/contacts", response_model=List[OppContactRoleOut])
def get_contact_roles(
    opp_id: str,
    ctx: dict = Depends(require_ui_session),
) -> List[OppContactRoleOut]:
    _require_flag()
    return list_contact_roles(ctx["user_sub"], opp_id)


@router.delete("/opportunities/{opp_id}/contacts/{contact_ref}", status_code=204, response_class=Response)
def delete_contact_role(
    opp_id: str,
    contact_ref: str,
    ctx: dict = Depends(require_ui_session),
):
    _require_flag()
    remove_contact_role(ctx["user_sub"], opp_id, contact_ref)


# ── OPP-005: Admin quota endpoints (on admin router) ─────────────────────────

@router_admin.post("/quotas", response_model=SalesQuotaOut, status_code=201)
async def create_quota(
    body: SalesQuotaIn,
    user: AuthenticatedUser = Depends(require_admin_or_root),
) -> SalesQuotaOut:
    _require_flag()
    return set_quota(user.sub, body)


@router_admin.get("/quotas/{user_sub}", response_model=SalesQuotaListOut)
async def list_user_quotas(
    user_sub: str,
    cursor: Optional[str] = Query(None),
    limit: int = Query(50, ge=1, le=200),
    user: AuthenticatedUser = Depends(require_admin_or_root),
) -> SalesQuotaListOut:
    _require_flag()
    limit = min(limit, 200)
    items, next_cursor = list_quotas_for_user(user_sub, cursor, limit)
    return SalesQuotaListOut(items=items, next_cursor=next_cursor)
